File: BinarySearchTrees/Java/BinarySearchTree.java
// BinarySearchTree.java - Implementation of a binary search tree

/**
 * Implementation of a generic binary search tree in Java.
 *
 * @author  Thomas Lang
 * @version 0.1
 */
public class BinarySearchTree<T extends Comparable<T>>{

    /**
     * Class representing a single Node of the tree.
     *
     * @author  Thomas Lang
     * @version 0.1
     */
    private class Node<T extends Comparable<T>>{
        private T value;
        private Node<T> left;
        private Node<T> right;
        private Node<T> parent;

        /**
         * Constructor, saves the passed generic value.
         *
         * @param  value   The value to save
         */
        private Node( T value ){
            this.value = value;
            left = right = parent = null;
        }

        /**
         * String representation of a node.
         *
         * @return  String representation
         */
        public String toString(){
            return ( "{" + value + "}" );
        }
    }


    // root node of the whole tree, we only have one
    // root node in a BINARY search tree
    Node<T> root;

    /**
     * Constructor, initializes the tree to null.
     */
    public BinarySearchTree(){ root = null; }

    /**
     * Adds a new node containing the value "value".
     *
     * @param  value   Value to insert.
     */
    public void add( T value ){
        if( this.isEmpty() ){
            root = new Node<T>( value );
        }else{
            insertInTree( root, value );
        }
    }

    /**
     * Inserts a new node containing the value "value"
     * to the binary search tree by using it's property
     * that all elements with a smaller value than the
     * actual node are left of this, the ones greater
     * then the actual node are right of it.
     *
     * @param  node   Actual node, for traversing
     * @param  value  Value to insert.
     */
    private void insertInTree( Node<T> node, T value ){
        if( value.compareTo( node.value ) < 0 ){
            if( node.left != null ){
                insertInTree( node.left, value );
            }else{
                node.left = new Node<T>( value );
                node.left.parent = node;
            }
        }else if( value.compareTo( node.value ) > 0 ){
            if( node.right != null )
                insertInTree( node.right, value );
            else{
                node.right = new Node<T>( value );
                node.right.parent = node;
            }
        }else{
            return;     // no duplicates allowed
        }
    }

    /**
     * Traverses the tree in Preorder.
     */
    public void traversePreorder(){
        if( root != null )
            traversePreorderRec( root );
        System.out.println();
    }

    /**
     * Traversing by first printing the current
     * node, then stepping through the left and
     * lastly through the right subtree.
     *
     * @param  node  Current node.
     */
    private void traversePreorderRec( Node<T> node ){
        System.out.print( node );
        if( node.left != null )
            traversePreorderRec( node.left );
        if( node.right != null )
            traversePreorderRec( node.right );
    }

    /**
     * Traverses the tree in Inorder direction.
     */
    public void traverseInorder(){
        if( root != null )
            traverseInorderRec( root );
        System.out.println();
    }

    /**
     * Traversing by first stepping throug the left
     * subtree of the current node, then printing it
     * and lastly stepping through the right subtree.
     *
     * @param  node  Current node.
     */
    private void traverseInorderRec( Node<T> node ){
        if( node.left != null )
            traverseInorderRec( node.left );
        System.out.print( node );
        if( node.right != null )
            traverseInorderRec( node.right );
    }

    /**
     * Traverses the tree in Postorder direction.
     */
    public void traversePostorder(){
        if( root != null )
            traversePostorderRec( root );
        System.out.println();
    }

    /**
     * Traversing by first stepping through the left 
     * and the right subtree of the current node and
     * lastly printing the current node.
     *
     * @param  node  Current node.
     */
    private void traversePostorderRec( Node<T> node ){
        if( node.left != null )
            traversePostorderRec( node.left );
        if( node.right != null )
            traversePostorderRec( node.right );
        System.out.print( node );
    }

    /**
     * Deletes the node containing the value "element" from the tree.
     *
     * @param  element  Value of the node-to-delete.
     */
    public void delete( T element ){
        if( isEmpty() )
            System.err.println( "Error! Cannot delete from empty Tree." );
        Node<T> nodeToDelete = findNode( element );
        if( nodeToDelete != null ){
            if( nodeToDelete.value.compareTo( root.value ) == 0 )
                deleteRoot();
            else if( nodeToDelete.left != null && nodeToDelete.right != null )
                deleteInternalNode( nodeToDelete );
            else 
                deleteNode( nodeToDelete );
        }else{
            System.err.println( "Error! Element not found" );
        }
    }

    /**
     * Deletes the root element.
     */
    private void deleteRoot(){
        deleteInternalNode( root );
    }

    /**
     * Deletes an internal node by swapping the 
     * node-to-delete with the next-smaller node
     * (which has either none or one child), and
     * then deleting this node.
     *
     * @param  node  Node-to-delete.
     */
    private void deleteInternalNode( Node<T> node ){
        // deleting internal node by swapping with
        // maximum of left child-subtree
        Node<T> maxOfMin  = findMaxOfMinNode( node );
        node.value = maxOfMin.value;
        deleteNode( maxOfMin );
    }

    /**
     * Deletes a node that has either one or no child.
     *
     * @param  node  The node to delete.
     */
    private void deleteNode( Node<T> node ){
        if(( node.left == null ) && ( node.right == null )){
            if( node.parent.right.value.compareTo( node.value ) == 0 ){
                node.parent.right = null;
                node.parent = null;
            }else{
                node.parent.left = null;
                node.parent = null;
            }
        }else if( node.left != null ){
            if( node.parent.right.value.compareTo( node.value ) == 0 ){
                node.parent.right = node.left;
                node.left.parent = node.parent;
                node.parent = null;
            }else{
                node.parent.left = node.left;
                node.left.parent = node.parent;
                node.parent = null;
            }
        }else{
            if( node.parent.right.value.compareTo( node.value ) == 0 ){
                node.parent.left = node.right;
                node.right.parent = node.parent;
                node.parent = null;
            }else{
                node.parent.right = node.right;
                node.right.parent = node.parent;
                node.parent = null;
            }
        }
    }

    /**
     * Finds a node containing the value "element" in the tree.
     *
     * @param   element  Value of the Node-to-Find
     * @return  The node with the value "element", null otherwise.
     */
    public Node<T> findNode( T element ){
        if( isEmpty() ) 
            System.err.println( "Error! No elements in Empty Tree." );
        else
            return findNodeRec( root, element );
        return null;
    }

    /**
     * Finds the node containing the value "element" by traversing
     * through the tree.
     *
     * @param  node     Current node.
     * @param  element  Value of the node-to-find.
     * @return The node, if found, null otherwise.
     */
    private Node<T> findNodeRec( Node<T> node, T element ){
        if( node.value.equals( element ) )
            return node;
        else if( node.value.compareTo( element ) < 0 )
            findNodeRec( node.left, element );
        else
            findNodeRec( node.right, element );
        return null;
    }

    /**
     * Finds the next-smaller node in the tree.
     *
     * @param  node  Current node.
     * @return next-smaller node in the tree.
     */
    public Node<T> findMaxOfMinNode( Node<T> node ){
        Node<T> leftChild = null;
        if( node.left != null ){
            leftChild = node.left;
            while( leftChild.right != null ){
                leftChild = leftChild.right;
            }
        }
        return leftChild;
    }

    /**
     * Checks whether the tree is empty or not.
     *
     * @return true if tree is empty, false otherwise
     */
    public boolean isEmpty(){ return ( root == null ); }

    /**
     * Prints the calling tree.
     */
    public void printTree(){
        if( isEmpty() )
            System.out.println( "( empty tree )" );
        else{
            System.out.println( "Root: " + root );
            printTreeRec( root );
        }
    }

    /**
     * Prints the tree recursively.
     *
     * @param  node  The current node.
     */
    private void printTreeRec( Node<T> node ){
        if( node.left != null )
            System.out.println( "Left: " + node.left );
        if( node.right != null )
            System.out.println( "Right: " + node.right );
        if( node.left != null )
            printTreeRec( node.left );
        if( node.right != null )
            printTreeRec( node.right );
    }
}

File: BinarySearchTrees/Python/binary_search_tree.py
#!/usr/env python
# 
# binary_search_tree.py - implementation of a binary search tree of Integers

class Node:
    """This class represents a single 
    Node of a binary search tree."""

    def __init__( self, value ):
        """Constructor, initializes the attributes."""
        self.value  = value
        self.bfsnum = 0
        self.left   = None
        self.right  = None
        self.parent = None

    def __str__( self ):
        """String representation of a single Node."""
        return "{" + str(self.value) + "}"


class BinarySearchTree:
    """This class represents a binary
    search tree and gives some methods/
    functions to operate with it."""

    def __init__( self ):
        """Constructor, initializes the root node to NULL."""
        self.root = None

    ##########################################################

    def isEmpty( self ):
        """Checks whether the tree is empty or not."""
        return ( self.root is None )

    ###########################################################

    def add( self, value ):
        """Adds a new Node containing the passed value
        to the tree by following the rules of a binary
        search tree.
        
        @param  value  The value to insert.
        """
        if self.isEmpty():
            self.root = Node( value )
        else:
            self.__insertIntoTree( self.root, value )


    def __insertIntoTree( self, node, value ):
        """Does the main inserting. Follows this rule:
                If the value is less than the current node,
                it goes into the left subtree. If it's greater
                then it steps into the right subtree.
                If none of the above things match, it will
                completely ignore it, because here are no 
                duplicates allowed.
        
        @param  node   Current node.
        @param  value  Value to insert.
            """
        if node.value > value:
            if node.left is not None:
                self.__insertIntoTree( node.left, value )
            else:
                node.left = Node( value )
                node.left.parent = node
        elif node.value < value:
            if node.right is not None:
                self.__insertIntoTree( node.right, value )
            else:
                node.right = Node( value )
                node.right.parent = node
        else:
            return      # no duplicates allowed

    ###########################################################

    def preorder( self ):
        """Traverses the tree in preorder direction."""
        if self.root is not None:
            self.__preorder( self.root )
        print


    def __preorder( self, node ):
        """The main traversing, goes recursively.
        
        @param  node  Current node.
        """
        print( node )
        if node.left is not None:
            self.__preorder( node.left )
        if node.right is not None:
            self.__preorder( node.right )

    ###########################################################

    def inorder( self ):
        """Traverses the tree in inorder direction."""
        if self.root is not None:
            self.__inorder( self.root )
        print


    def __inorder( self, node ):
        """The main traversing, goes recursively
        
        @param  node  Current node.
        """
        if node.left is not None:
            self.__inorder( node.left )
        print( node )
        if node.right is not None:
            self.__inorder( node.right )

    ###########################################################

    def postorder( self ):
        """Traverses the tree in postorder direction."""
        if self.root is not None:
            self.__postorder( self.root )
        print


    def __postorder( self, node ):
        """The main traversing, goes recursively.
        
        @param  node  Current node.
        """
        if node.left is not None:
            self.__postorder( node.left )
        if node.right is not None:
            self.__postorder( node.right )
        print( node )

    ###########################################################

    def breadthfirst( self ):
        """Traverses the tree in breadthfirst-direction.
        Normally we would use a Queue, but instead of 
        importing it we just simulate it with a list."""
        queue = []
        queue.append( self.root )
        bfscount = 1
        self.root.bfsnum = bfscount
        while( len( queue ) != 0 ):
            v = queue.pop()
            print( v )
            bfscount += 1; v.bfsnum = bfscount
            if v.left is not None:
                queue.insert( 0, v.left )
            if v.right is not None:
                queue.insert( 0, v.right )
        print

    ###########################################################

    def delete( self, value ):
        """Deletes the Node containing 'value' from the tree.
        
        @param  value  Content of the node-to-delete
        """
        if self.isEmpty():
            raise ValueError( "Error! Cannot delete from empty tree." )
        
        nodeToDelete = self.__findNode( value )
        
        if nodeToDelete is not None:
            if nodeToDelete.value == self.root.value:
                self.__deleteInternalNode( self.root )
            elif ( nodeToDelete.left is not None ) and ( nodeToDelete.right is not None ):
                self.__deleteInternalNode( nodeToDelete )
            else:
                self.__deleteNode( nodeToDelete )
        else:
            raise ValueError( "Error! Element not found." )


    def __deleteInternalNode( self, node ):
        """Deletes an internal node by swapping it's value
        with the next-less node and then deleting that node.
        
        @param  node  Current node
        """
        maxOfMin = self.__findMaxOfMinNode( node )
        node.value = maxOfMin.value
        self.__deleteNode( maxOfMin )


    def __deleteNode( self, node ):
        """Deletes any other node and looks, if it's a left 
        or a right child (to guarantee after the process that
        we also have a search tree.
        
        @param  node  Current node.
        """
        if ( node.left is None ) and ( node.right is None ):
            if node.parent.right.value == node.value:
                node.parent.right = None
                node.parent = None
            else:
                node.parent.left = None
                node.parent = None
        elif node.left is not None:
            if node.parent.right.value == node.value:
                node.parent.right = node.left
                node.left.parent = node.parent
                node.parent = None
            else:
                node.parent.left = node.left
                node.left.parent = node.parent
                node.parent = None
        else:
            if node.parent.right.value == node.value:
                node.parent.left = node.right
                node.right.parent = node.parent
                node.parent = None
            else:
                node.parent.right = node.right
                node.right.parent = node.parent
                node.parent = None


    def __findNode( self, element ):
        """Finds a specific node containing the value
        'element' in the tree."""
        if self.isEmpty():
            print( "Error! No elements in empty tree." )
        else:
            return self.__findNodeRec( self.root, element )
        

    def __findNodeRec( self, node, element ):
        """Does the main finding, goes recursively 
        through the tree"""
        if node.value == element:
            return node
        elif node.value > element:
            return self.__findNodeRec( node.left, element )
        else:
            return self.__findNodeRec( node.right, element )


    def __findMaxOfMinNode( self, node ):
        """Finds the maximum node of the left
        subtree of node, which is needed for
        deleting.
        
        @param  node  Current node
        """
        if node.left is not None:
            leftChild = node.left
            while leftChild.right is not None:
                leftChild = leftChild.right
        return leftChild 


    ###########################################################

    def __treeRep( self, node ):
        """Internal String representation. This is
        constructed by stepping recursively through 
        the tree in preorder direction.
        
        @param  node  Current node.
        """
        out = "Node:" + str(node) + "\n"
        if node.left is not None:
            out += " Left: " + str(node.left)
        if node.right is not None:
            out += " Right: " + str(node.right)
        out += "\n"
        if node.left is not None:
            out += self.__treeRep( node.left )
        if node.right is not None:
            out += self.__treeRep( node.right )
        return out

    def __str__( self ):
        """String representation of a binary search tree."""
        output = ""
        if self.root is not None:
            output += self.__treeRep( self.root )
        else:
            output += "(empty tree)"
        return output

################################################################

def main():
    """Main method, does some testing."""
    print( "*** Initializing new binary search tree ..."    )
    print( "*** Adding values: 5, 7, 9, 1, 4, 0, 6, 2, 42 " )
    t = BinarySearchTree()
    t.add( 5 ); t.add( 7 ); t.add( 9 ); t.add( 1 );
    t.add( 4 ); t.add( 0 ); t.add( 6 ); t.add( 2 ); t.add( 42 );
    print( "\n  ### Actual tree: \n" )
    print( t )
    print( "Preorder-traversal:" )
    t.preorder()
    print( "Inorder-traversal: ")
    t.inorder()
    print( "Postorder-traversal: ")
    t.postorder()
    print( "Breadth-first-traversal: " )
    t.breadthfirst()
    print; print
    print( "Deleting Node containing \"42\" ..." )
    t.delete( 42 )
    print( "\n  ### Actual tree: \n" )
    print( t )
    print; print
    print( "Deleting Node containing \"5\" ..." )
    t.delete( 5 )
    print( "\n  ### Actual tree: \n" )
    print( t )

if __name__ == "__main__":
    main()
